$OpenBSD: patch-test_static_test_rb,v 1.1.2.1 2009/06/16 02:02:13 william Exp $

Security fix for directory traversal vulnerability in static file route [#177]

--- test/static_test.rb.orig	Fri Jun 12 12:31:59 2009
+++ test/static_test.rb	Fri Jun 12 12:32:58 2009
@@ -64,4 +64,19 @@ describe 'Static' do
     get "/foobarbaz.txt"
     assert not_found?
   end
+
+  it 'serves files when .. path traverses within public directory' do
+    get "/data/../#{File.basename(__FILE__)}"
+    assert ok?
+    assert_equal File.read(__FILE__), body
+  end
+
+  it '404s when .. path traverses outside of public directory' do
+    mock_app {
+      set :static, true
+      set :public, File.dirname(__FILE__) + '/data'
+    }
+    get "/../#{File.basename(__FILE__)}"
+    assert not_found?
+  end
 end
